File: interface_example.java
interface PaymentMethod {
    void pay(int amount);
}
class CardPayment2 implements PaymentMethod {
    public void pay(int amount) {
//logic
    }
}

class UpiPayment implements PaymentMethod {
    public void pay(int amount) {
//logic    }
}

class CheckoutService {
    private final PaymentMethod payment;

    CheckoutService(PaymentMethod payment) {
        this.payment = payment;
    }

    void checkout(int amount) {
        payment.pay(amount); // polymorphism
    }
}

File: interface_example.cpp
//
// Created by nived on 09-01-2026.
//
#include <iostream>
using namespace std;

class PaymentMethod {
public:
    virtual void pay(int amount) = 0;
    virtual ~PaymentMethod() = default;
};

class CardPayment : public PaymentMethod {
public:
    void pay(int amount) override {
        cout << "Card payment: " << amount << endl;
    }
};

class CheckoutService {
    PaymentMethod& payment;
public:
    CheckoutService(PaymentMethod& p) : payment(p) {}

    void checkout(int amount) {
        payment.pay(amount); // dynamic dispatch
    }
};
